### PR TITLE
Reorganize Demo App & Add rxExample support

### PR DESCRIPTION
**Pull request by CITguy**
Reorganized demo directory to simplify development and make it easier for new devs to find code. As such, it was reorganized as if it were a separate application (which it technically is).

New demo directory structure is as follows:

* `demo/`
  * `bower_components/` (copied to `build/bower_components`)
  * `css/` (copied to `build/css`)
  * `images/` (copied to `build/images`)
  * `index.html` (copied to `build/indx.html`)
  * `less/` (gets compiled into `build/css`)
  * `scripts/` (copied/compiled to `build/scripts`)
  * `templates/` (copied/compiled to `build/templates`)

To support new, demo-only `<rx-example>` directive functionality, the module directory structure now supports the following (*note the `docs/examples` directory*)

* `src/**/<module>/`
  * `docs/`
    * `<module>.html`
    * `<module>.js`
    * `<module>.demo.less` *(optional)* (no more need to define demo-only styles within the HTML)
    * `examples/`
      * `<module>.<exampleName>.html`
      * `<module>.<exampleName>.js` *(optional)*
      * `<module>.<exampleName>.less` *(optional)*
  * `README.md` (it is strongly recommended that content be moved to `docs/<module>.html`)
  * `<module>.module.js` (supported for legacy components, unnecessary for quarks, atoms, and molecules)
  * `<module>.meta.json` (**required for all modules**)
  * `<module>.less` (business as usual)
  * `scripts/` (business as usual)
  * `images/` (business as usual)
  * `templates/` (business as usual)

### Questions
- Will the gh-pages build task check in `build/css/demo.css` (generated from `demo/less/demo.less`)?
  - **Yes, anything in the `build/` directory gets committed to the `gh-pages` branch**

### JIRA
* [FRMW-367](https://jira.rax.io/browse/FRMW-367)
* [FRMW-368](https://jira.rax.io/browse/FRMW-368)
* [FRMW-419](https://jira.rax.io/browse/FRMW-419)

### LGTMs
- [x] Dev LGTM
- [x] Dev LGTM
- [x] Design LGTM

### CODE REVIEW

File: demo/modules.js [hunk 1]
```diff
@@ -1,3 +0,0 @@
-angular.module('demoApp')
-.value('components', <%= JSON.stringify(config.components, null, 4) %>)
```
Comment (CITguy) [inline demo/modules.js]: no longer necessary

File: demo/modules.js [hunk 2]
```diff
@@ -1,3 +0,0 @@
...
-.value('modules', <%= JSON.stringify(config.modules, null, 4) %>);
```
Comment (CITguy) [inline demo/modules.js]: moved to `demo/scripts/loadModules.js`

File: demo/guides/guideViewer.html [hunk 3]
```diff
@@ -1,3 +0,0 @@
-<rx-page title="vm.title">
-    <div class="markdown clear" ng-include="'guides/rendered/' + vm.guide + '.html'" onload="vm.activate()"></div>
-</rx-page>
```
Comment (CITguy) [inline demo/guides/guideViewer.html]: moved to `demo/templates/guides/guideViewer.html`

File: demo/componentsController.js [hunk 4]
```diff
@@ -1,65 +0,0 @@
...
-        function showCaret (key) {
-            return vm.sortKey === key;
-        }
-
-        function sortColumn (key) {
-            // Click new column defaults to ascending; click same column toggles between asc and desc
-            vm.sortReverse = (key === vm.sortKey ? !vm.sortReverse : false);
-            vm.sortKey = key;
-        }
-    });
```
Comment (CITguy) [inline demo/componentsController.js]: replaced by `listModulesController` and `listCategoryModulesController`

File: demo/scripts/demoApp.js [hunk 5]
```diff
@@ -18,145 +18,235 @@ function genericRouteController (breadcrumbs) {
...
-.controller('componentCtrl', function ($scope, rxBreadcrumbsSvc, $routeParams, component) {
-    rxBreadcrumbsSvc.set([{
-            path: '#/components',
-            name: 'Components'
-        }, {
-            name: component.name
-        }]);
-
-    $scope.component = component;
-})
```
Comment (CITguy) [inline demo/scripts/demoApp.js]: moved to `demo/scripts/showModuleController.js`

File: demo/scripts/demoApp.js [hunk 6]
```diff
@@ -362,56 +470,6 @@ angular.module('demoApp', ['encore.ui', 'ngRoute'])
         }
     });
 })
-.directive('rxPrism', function ($timeout, $http) {
```
Comment (CITguy) [inline demo/scripts/demoApp.js]: moved to `demo/scripts/rxPrism.js`

File: demo/scripts/demoApp.js [hunk 7]
```diff
@@ -362,56 +470,6 @@ angular.module('demoApp', ['encore.ui', 'ngRoute'])
...
-                    });
-                } else {
-                    Prism.highlightElement(code);
-                }
-            }, 0);
-        }
-    };
-})
-
-.directive('alwaysInvalid', function () {
```
Comment (CITguy) [inline demo/scripts/demoApp.js]: moved to `demo/scripts/alwaysInvalid.js`

File: demo/scripts/demoApp.js [hunk 8]
```diff
@@ -362,56 +470,6 @@ angular.module('demoApp', ['encore.ui', 'ngRoute'])
...
-            /* ng1.2+ (hack) */
-            el.addClass('ng-dirty');
-            /* ng1.3+ */
-            //ctrl.$setDirty();
-        }
-    };
-})
-
-// Used for drawing the Demo and Markup tabs in the styleguide
-.directive('rxStyleguide', function () {
```
Comment (CITguy) [inline demo/scripts/demoApp.js]: moved to `demo/scripts/rxStyleguide.js`

File: demo/scripts/listModulesController.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+ * @description
+ * Used to filter modules list. Given a string, filter checks for the following case insensitive conditions:
+ *  - input string matches any part of module name value
+ *  - input string matches begining of stability value
+ *  - input string matches any part of module category
+ *
+ * @param {Array} modules list of modules to be filtered
+ * @param {String} search input string to filter list by
+ */
+.filter('modulesFilter', function () {
```
Comment (CITguy) [inline demo/scripts/listModulesController.js]: note to self: move to `demo/scripts/modulesFilter.js`

File: demo/scripts/listCategoryModulesController.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ * @name demoApp:categoryModulesFilter
+ * @description
+ * Used to filter module list. Given a string, filter checks for the following case insensitive conditions:
+ *  - input string matches any part of module name value
+ *  - input string matches begining of stability value
+ *
+ * @param {Array} modules list of modules to be filtered
+ * @param {String} search input string to filter list by
+ */
+.filter('categoryModulesFilter', function () {
```
Comment (CITguy) [inline demo/scripts/listCategoryModulesController.js]: note to self: move to `demo/scripts/categoryModulesFilter.js`

File: demo/scripts/showModuleController.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+angular.module('demoApp')
+.controller('showModuleController', function ($scope, $filter, rxBreadcrumbsSvc, $routeParams, module) {
+    rxBreadcrumbsSvc.set([
+        {
+            path: '#/' + module.category,
+            name: $filter('rxCapitalize')(module.category)
```
Comment (CITguy) [inline demo/scripts/showModuleController.js]: breadcrumb is dynamic based on the category of the module being viewed

File: demo/styleguide/basics/lists.html [hunk 12]
```diff
@@ -1,37 +0,0 @@
-<div class="pure-g columns clear">
```
Comment (CITguy) [inline demo/styleguide/basics/lists.html]: File no longer used. Also, it was using deprecated pure classes.

File: demo/templates/modules/listCategoryModules.html [hunk 13]
```diff
@@ -0,0 +1,40 @@
+<rx-page title="vm.pageTitle" class="modules-list">
+  <div class="clear">
+  </div>
```
Comment (CITguy) [inline demo/templates/modules/listCategoryModules.html]: note to self: remove empty div

File: demo/templates/modules/listModules.html [hunk 14]
```diff
@@ -0,0 +1,44 @@
+<rx-page title="vm.pageTitle" class="modules-list">
+  <div class="clear">
+  </div>
```
Comment (CITguy) [inline demo/templates/modules/listModules.html]: note to self: remove empty div

File: demo/templates/modules/showModule.html [hunk 15]
```diff
@@ -0,0 +1,40 @@
+<rx-page title="module.name">
+  <div layout="row" layout-align="middle">
+    <div flex>
+      <img ng-if="module.stability"
+        alt="stability is {{module.stability}}"
+        ng-src="http://badges.github.io/stability-badges/dist/{{module.stability}}.svg" />
+    </div>
+    <div flex class="align-right">
+      <a href="ngdocs/index.html#/api/{{ module.name }}/" class="module-api button">View API</a>
+      <a href="https://github.com/rackerlabs/encore-ui/tree/master/src/{{ module.category }}/{{ module.name }}/" class="module-source button">View Source</a>
```
Comment (CITguy) [inline demo/templates/modules/showModule.html]: href updated to reflect new `src/` directory structure by adding `/{{ module.category }}` to the path

File: src/quarks/hotkeys/hotkeys.module.js [hunk 16]
```diff
@@ -45,4 +45,3 @@
  * help window.
  *
  */
```
Comment (CITguy) [inline src/quarks/hotkeys/hotkeys.module.js]: moved `hotkeys` to `src/quarks/` as an example and to test out build tasks, etc.

File: src/quarks/hotkeys/hotkeys.module.js [hunk 17]
```diff
@@ -45,4 +45,3 @@
...
-angular.module('encore.ui.hotkeys', []);
```
Comment (CITguy) [inline src/quarks/hotkeys/hotkeys.module.js]: hotkeys needs no module definition. It has no directly executable code, and all quarks will be defined under the `encore.ui.quarks` namespace (simplifying number of required modules)

File: demo/templates/modules/showModule.html (hunk 15 above)
Comment (100stacks) [inline demo/templates/modules/showModule.html]: Any gains by storing these stability badges locally? 

> ng-src="http://badges.github.io/stability-badges/dist/{{module.stability}}.svg" />

Comment (CITguy) [inline demo/templates/modules/showModule.html]: not really

**Comment by parlarjb** [comment]
This is awesome!

**Comment by 100stacks** [comment]
Tame the beast! :japanese_ogre:

File: .gitignore [hunk 18]
```diff
@@ -34,6 +34,7 @@ results
 /templates/
 /index.js
 /build/
+/tmp/
```
Comment (CITguy) [inline .gitignore]: moved `build/tmp` to project root to exclude from gh-pages commit

File: demo/assets/generated_demo.css [hunk 19]
```diff
@@ -1,375 +0,0 @@
-/* Code Styles */
```
Comment (CITguy) [inline demo/assets/generated_demo.css]: File is no longer necessary. `demo/less/demoApp.less` (previously `demo/assets/demo.less`) compiles to `build/css/demoApp.css` (which is committed to the `gh-pages` branch). It doesn't need committed to `master` or any feature branch.

File: demo/component-template.html [hunk 20]
```diff
@@ -1,40 +0,0 @@
-<rx-page title="component.name">
```
Comment (CITguy) [inline demo/component-template.html]: Replaced by `demo/templates/modules/showModule.html` as part of demo app reorganization.

File: demo/components.html [hunk 21]
```diff
@@ -1,35 +0,0 @@
-<rx-page title="'All Components'" class="components-list">
```
Comment (CITguy) [inline demo/components.html]: Replaced by `demo/templates/modules/listModules.html` and `demo/templates/modules/listCategoryModules.html` as part of demo app reorganization.

File: demo/componentsController.js [hunk 22]
```diff
@@ -1,65 +0,0 @@
-angular
```
Comment (CITguy) [inline demo/componentsController.js]: Replaced by `demo/scripts/listModulesController.js` and `demo/scripts/listCategoryModulesController.js`

File: demo/index.html [hunk 23]
```diff
@@ -9,11 +10,11 @@
...
 
-    <link href="dist/<%= config.fileNameResp %>.css" rel="stylesheet" />
-    <link href="assets/prism.css" rel="stylesheet" />
-    <link href="assets/<%= config.styleguideCssFilename %>" rel="stylesheet" />
-
-    <script src="assets/prism.js"></script>
+    <link href="dist/<%= config.dist.fileNameResp %>.css" rel="stylesheet" />
+    <link href="css/prism.css" rel="stylesheet" />
+    <link href="<%= config.css.demoApp %>" rel="stylesheet" />
+    <link href="<%= config.css.demos %>" rel="stylesheet" />
```
Comment (CITguy) [inline demo/index.html]: This stylesheet contains demo-only styles as defined in `src/**/docs/*.demo.less`

File: demo/index.html [hunk 24]
```diff
@@ -9,11 +10,11 @@
...
+    <link href="<%= config.css.examples %>" rel="stylesheet" />
```
Comment (CITguy) [inline demo/index.html]: This stylesheet contains example-specific styles as defined in `src/**/docs/examples/*.less`

File: demo/scripts/alwaysInvalid.js [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+        require: 'ngModel',
+        link: function (scope, el, attrs, ctrl) {
+            ctrl.$setValidity('alwaysInvalid', false);
+            /* ng1.2+ (hack) */
+            el.addClass('ng-dirty');
+            /* ng1.3+ */
+            //ctrl.$setDirty();
+        }
+    };
+});
```
Comment (CITguy) [inline demo/scripts/alwaysInvalid.js]: extracted from `demo/scripts/demoApp.js`

File: demo/scripts/demoApp.js [hunk 26]
```diff
@@ -18,145 +18,229 @@ function genericRouteController (breadcrumbs) {
...
-        .when('/components', {
-            templateUrl: 'components.html',
-            controller: 'componentsController as vm'
+        /* Module Pages */
+        .when('/modules', {
+            templateUrl: 'templates/modules/listModules.html',
+            controller: 'listModulesController',
+            controllerAs: 'vm'
         })
-        .when('/component/:component', {
```
Comment (CITguy) [inline demo/scripts/demoApp.js]: Redirect removed. All known locations within documentation have been updated to point to `#/components/:component`

File: demo/scripts/rxPrism.js [hunk 27]
```diff
@@ -0,0 +1,27 @@
+angular.module('demoApp')
```
Comment (CITguy) [inline demo/scripts/rxPrism.js]: Extracted from `demo/scripts/demoApp.js`

File: demo/scripts/rxStyleguide.js [hunk 28]
```diff
@@ -0,0 +1,16 @@
+/**
```
Comment (CITguy) [inline demo/scripts/rxStyleguide.js]: Extracted from `demo/scripts/demoApp.js`

File: grunt-tasks/build.js [hunk 29]
```diff
@@ -5,8 +5,9 @@ module.exports = function (grunt) {
             'modules',
             'concat:dist',
             'concat:distTpls',
-            'concat:tmpLess',
-            'concat:tmpLessResp',
+            'concat:tmpEncoreLess',
```
Comment (CITguy) [inline grunt-tasks/build.js]: `concat:tmpEncoreLess` handles logic for both `concat:tmpLess` and `concat:tmpLessResp`

File: grunt-tasks/modules.js [hunk 30]
```diff
@@ -123,7 +132,8 @@ module.exports = function (grunt) {
 
         //- used in grunt-tasks/phantomjs-checks.js
         //  - run via `grunt test:full` on TravisCI
-        grunt.config('config.demoModules', grunt.config('config.components'));
+        //TODO: phantomjs-checks and this config will likely need updated as Atoms and Molecules are added.
+        grunt.config('config.demoModules', _.filter(modules, { category: 'components' }));
```
Comment (CITguy) [inline grunt-tasks/modules.js]: *note to self*: update for compatibility with phantomjs-checks task

File: grunt-tasks/options/copy.js [hunk 31]
```diff
@@ -77,48 +77,81 @@ function convertMarkdown (content) {
...
+            src: [
+                'css/*.css',
+                'images/**/*',
+                'bower_components/**/*'
+            ],
+            dest: '<%= config.dir.docs %>'
         }]
     },
+    /* Copy polyfils for demo app. */
     demopolyfills: {
```
Comment (CITguy) [inline grunt-tasks/options/copy.js]: *note to self* is this task needed?

**Comment by CITguy** [comment]
## All Modules
0. Sort by Category (clicking value in cell will navigate to appropriate category listing)
0. Search across Atoms, Components, Molecules, **and** Quarks in one place.
all-modules

**Comment by CITguy** [comment]
## All Components
0. Link will navigate to "All Modules" page, if devs have `#/components` bookmarked.
0. "Modules" breadcrumb also navigates to `"All Modules" page.
all-components

**Comment by CITguy** [comment]
## hotkeys quark
quarks-hotkeys

1. Breadcrumbs provides ability to navigate to "All Quarks" or "All Modules" listing.
2. "All Modules" link in left-hand have allows user to search across Atoms, Components, Molecules, **and** Quarks in one place.
3. Sub navigation links lead to the appropriate category listing. Further sub navigation was left out as it would introduce unnecessary complexity for this already complex PR.
4. `hotkeys` documentation uses new `<rx-example>` directive to display demo side-by-side with its source code. Additionally, support is available for demo to be above source code. This example can be included anywhere in the demo app or module documentation and you only need to keep one source location updated.

**Comment by 100stacks** [comment]
@CITguy great work on restructuring/updating the build script and demo app! LGTM :+1:

**Comment by glynnis** [comment]
This LGTM on the design side, but I'mma page @lfielder to give this a once-over since he's been more involved in the docs redesign/IA work.

This is :angel: :angel: :sparkles: :rainbow: :tada:. :tophat:s off to @CITguy

**Comment by glynnis** [comment]
Already chatted some with @CITguy about this, but wanted to capture somewhere in writing: it's not relevant to this PR (should be dealt with in a separate story), but one question I still have about this structure and future IA work is how we will deal with `demo/` vs. `src/` examples and pages. Today we have a division between the "styleguide" (`demo/`) and component pages (`src/`), which means that some examples are duplicated and are frequently inconsistent as a result. Do we know whether/how we want to address that moving forward? cc/ @lfielder @parlarjb

**Comment by CITguy** [comment]
### Quark - Hotkeys
quarks-hotkeys-2

After chatting with @lfielder.

1. `<rx-example>` updated to use tabbed layout for familiarity and simplicity until a better design for examples can be developed.
2. It was desired to have accordion-like navigation for Atoms, Molecules, etc. to show sublinks, however limitations of the nav menu left us with this less desired structure.

**Comment by CITguy** [comment]
@glynnis, the `<rx-example>` support added with this PR will help by providing a single place for example code so the examples can be used anywhere (in `demo/` or `src/`). An example would be the `rxButton.sizing` example that is likely to appear in the near future. You can place the button sizing example in as many document locations as desired, and you'd only need to update the example in one location, should it need updated.  This does not, however, prevent overuse/duplication of examples and your point of "demo" vs "styleguide" is still relevant.

File: demo/assets/generated_demo.css (hunk 19 above)
Comment (parlarjb) [inline demo/assets/generated_demo.css]: ♪┏(・o･)┛♪┗ ( ･o･) ┓♪

**Comment by parlarjb** [comment]
LGTM

**Pull request merged**